src/FeedGenerator.php

```diff
@@ -101,9 +104,115 @@ function () {
+		// PHP shuts down execution for some reason.
+		add_action( 'action_scheduler_unexpected_shutdown', array( $this, 'handle_unexpected_shutdown' ), 10, 2 );
+		// Action got an exception thrown.
+		add_action( 'action_scheduler_failed_execution', array( $this, 'handle_failed_execution' ), 10, 3 );
+	}
+
+	/**
+	 * Returns feed generator actions group name.
+	 *
+	 * @since x.x.x
+	 *
+	 * @return string
+	 */
+	public function get_group_name(): string {
+		return PINTEREST_FOR_WOOCOMMERCE_PREFIX;
+	}
+
+	/**
+	 * Unexpected shutdown handler.
+	 *
+	 * @param int        $action_id - Action Scheduler action ID.
+	 * @param array|null $error - Error details.
+	 *
+	 * @since x.x.x
+	 *
+	 * @return void
+	 */
+	public function handle_unexpected_shutdown( int $action_id, ?array $error ) {
+		if ( ! $error || ! $this->is_timeout_error( $error ) ) {
+			return;
+		}
+		$action = ActionScheduler::store()->fetch_action( $action_id );
+		$hook   = $action->get_hook();
+		$args   = $action->get_args();
+
+		// If not Pinterest Feed Generator action - ignore.
+		if ( $hook !== $this->get_action_full_name( self::CHAIN_BATCH ) ) {
+			return;
+		}
+
+		// Check if the action had failed before.
+		if ( $this->is_failure_rate_above_threshold( $hook, $args ) ) {
+			self::log(
+				sprintf(
+					// Translators: 1. Action Scheduler hook name.
+					__(
+						'Feed Generator `%s` Action reschedule threshold has been reached. Quit.',
+						'pinterest-for-woocommerce'
+					),
+					$hook
+				)
+			);
+			return;
+		}
+
+		self::log(
+			sprintf(
+				// Translators: Action Scheduler hook name.
+				__(
+					'Feed Generator `%s` Action timed out due to an unexpected shutdown. Rescheduling it.',
+					'pinterest-for-woocommerce'
+				),
+				$hook
+			)
+		);
+
+		// Decrease the number of products to retry.
+		$attempt = ( Pinterest_For_Woocommerce::get_data( 'feed_product_batch_attempt' ) ?? 1 ) + 1;
+		$limit   = (int) ceil( $this->get_batch_size() / $attempt );
+		Pinterest_For_Woocommerce::save_data( 'feed_product_batch_size', $limit );
+		Pinterest_For_Woocommerce::save_data( 'feed_product_batch_attempt', $attempt );
+
+		self::log(
+			sprintf(
+				// Translators: 1: Action Scheduler hook name, 2: New products number to process next action run.
+				__(
+					'Feed Generator `%1$s` Action product batch size decreased to %2$d.',
+					'pinterest-for-woocommerce'
+				),
+				$hook,
+				$limit
+			)
+		);
+
+		// Register retry attempt.
+		$this->action_scheduler->schedule_immediate( $hook, $args, PINTEREST_FOR_WOOCOMMERCE_PREFIX );
+	}
+
+	/**
+	 * Action exception handler.
+	 *
+	 * @param int       $action_id - Action Scheduler action id.
+	 * @param Throwable $throwable - Exception object.
+	 * @param string    $context - Action Scheduler context (e.g. WP CLI, Async Request or an empty string).
+	 *
+	 * @since x.x.x
+	 *
+	 * @return void
+	 */
+	public function handle_failed_execution( int $action_id, Throwable $throwable, string $context ) {
```

Comment:
The `$context` parameter doesn't seem to be used here.

Reply:
Removed the parameter from the list and updated corresponding actions.